AX: content of <script> elements announced by VoiceOver inside containers with display: contents

twilco opened the pull request:
#### db2c73f6239bc40f2d53b4d3b332a7ee531897ca
<pre>
AX: content of &lt;script&gt; elements announced by VoiceOver inside containers with display: contents
<a href="https://bugs.webkit.org/show_bug.cgi?id=283645">https://bugs.webkit.org/show_bug.cgi?id=283645</a>
<a href="https://rdar.apple.com/problem/140493798">rdar://problem/140493798</a>

Reviewed by Chris Fleizach.

Up until this point, we have got away with not handling display:none (which is the user-agent default `display` for
script) because it causes the content within to not have a renderer, and thus not be a candidate for AX object creation,
minus a few exceptions — display:contents, elements within a canvas, a few others, and most notably: elements descending
from another element involved in a relation.

Because the unrendered text inside the script tag is descendant of an element with a relation, we create an AX object
for it. And because AccessibilityNodeObject::computeIsIgnored didn&apos;t check for a display:none ancestor, we compute
the unrendered text inside the script to be unignored.

With this commit, we check for display:none in our existing walk up the ancestry as part of the
AccessibilityObject::defaultObjectInclusion(), and ignore the object if we find one.

* LayoutTests/accessibility/display-none-descendant-of-relation-expected.txt: Added.
* LayoutTests/accessibility/display-none-descendant-of-relation.html: Added.
* LayoutTests/platform/glib/TestExpectations: Skip new test.
* LayoutTests/platform/ios/TestExpectations: Enable new test.
* LayoutTests/platform/ios/accessibility/display-none-descendant-of-relation-expected.txt: Added.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::isDOMHidden):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::computeIsIgnored const):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::style const):
(WebCore::AccessibilityObject::existingStyle const):
(WebCore::AccessibilityObject::defaultObjectInclusion const):
* Source/WebCore/accessibility/AccessibilityObject.h:

Canonical link: <a href="https://commits.webkit.org/287159@main">https://commits.webkit.org/287159@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/234a7b89bd1e0c7743525bf4ed0727e1523e9abd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78555 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57600 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31937 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83216 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29820 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/80688 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66751 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5881 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61532 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19449 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81622 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51542 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/69840 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41842 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48888 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/25384 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28157 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69988 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/25757 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84582 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5920 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4057 "Found 1 new test failure: imported/w3c/web-platform-tests/html/browsers/the-window-object/open-close/open-features-tokenization-noreferrer.html (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69756 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6081 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67519 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69010 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17198 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13030 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/11483 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5867 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/11828 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5855 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9289 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7642 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fleizach commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -4814,9 +4814,14 @@ bool isNodeFocused(Node& node)
+    return style.display() == DisplayType::None || style.usedVisibility() != Visibility::Visible;
+}
+
 bool isDOMHidden(const RenderStyle* style)
```
Is this DOM hidden? Or is it RenderHidden?

twilco replied:
Yeah, `isRenderHidden` is probably more accurate here. I've renamed the function to that in the latest commit, thanks!

twilco committed:
AX: content of <script> elements announced by VoiceOver inside containers with display: contents
https://bugs.webkit.org/show_bug.cgi?id=283645
rdar://problem/140493798

Reviewed by Chris Fleizach.

Up until this point, we have got away with not handling display:none (which is the user-agent default `display` for
script) because it causes the content within to not have a renderer, and thus not be a candidate for AX object creation,
minus a few exceptions — display:contents, elements within a canvas, a few others, and most notably: elements descending
from another element involved in a relation.

Because the unrendered text inside the script tag is descendant of an element with a relation, we create an AX object
for it. And because AccessibilityNodeObject::computeIsIgnored didn't check for a display:none ancestor, we compute
the unrendered text inside the script to be unignored.

With this commit, we check for display:none in our existing walk up the ancestry as part of the
AccessibilityObject::defaultObjectInclusion(), and ignore the object if we find one.

* LayoutTests/accessibility/display-none-descendant-of-relation-expected.txt: Added.
* LayoutTests/accessibility/display-none-descendant-of-relation.html: Added.
* LayoutTests/platform/glib/TestExpectations: Skip new test.
* LayoutTests/platform/ios/TestExpectations: Enable new test.
* LayoutTests/platform/ios/accessibility/display-none-descendant-of-relation-expected.txt: Added.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::isDOMHidden):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::computeIsIgnored const):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::style const):
(WebCore::AccessibilityObject::existingStyle const):
(WebCore::AccessibilityObject::defaultObjectInclusion const):
* Source/WebCore/accessibility/AccessibilityObject.h:

Canonical link: https://commits.webkit.org/287159@main
